flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ExecutionDecorator.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.streaming.runtime.tasks.mailbox.Mail;
+import org.apache.flink.util.function.ThrowingRunnable;
+
+import java.util.concurrent.Callable;
+
+/**
+ * Wraps execution of a {@link Runnable}, {@link ThrowingRunnable}, {@link Callable}, or {@link Mail}.
+ * Intended to customize execution in sub-types fo {@link org.apache.flink.streaming.runtime.tasks.StreamTask StreamTask},
+ * e.g. synchronization in {@link org.apache.flink.streaming.runtime.tasks.SourceStreamTask SourceStreamTask}.
+ */
+public interface ExecutionDecorator {
```

Comment:
I'd prefer to have a name that doesn't include the name of the design pattern. Could we name it (Work)Executor or (Work)Synchronizer?

Reply:
Add `@Internal`.

Reply:
> doesn't include the name of the design pattern

Why not? Having it in the name makes it's purpose clear straight away.
Executor and Synchronizer are incorrect, as it nor executes, nor necessarily synchronizes.

Reply:
Hm why isn't it executing? It's executing runnables, mails, etc. ;)
Btw, my main problem with ExecutorDecorator is that it's actually not Decorating anything. So it's a misnomer.

Reply:
To me, `Executor` is usually related to threads (e.g. in j.u.c.). While same-thread implementation might be available, it is usually not the intent of creating the interface. Here the intent is to decorate each execution. And both design pattern and interface name should reflect the purpose of the abstraction.

But it looks like we won't reach consensus in a foreseeable future :) So I renamed it to `StreamTaskActionExecutor` and its ex-`NO_OP` implementation to `ImmediateExecStreamTaskActionExecutor`.